webapp/packages/plugin-navigation-tree/src/TreeNew/useTreeVirtualization.ts

```diff
@@ -0,0 +1,87 @@
+/*
+ * CloudBeaver - Cloud Database Manager
+ * Copyright (C) 2020-2024 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0.
+ * you may not use this file except in compliance with the License.
+ */
+import { observable, runInAction } from 'mobx';
+import { useEffect } from 'react';
+
+import { useObjectRef } from '@cloudbeaver/core-blocks';
+
+interface IPrivate extends ITreeVirtualization {
+  observer: ResizeObserver | null;
+  element: HTMLElement | null;
+  handleScroll(event: Event): void;
+  handleResize(): void;
+  dispose(): void;
+}
+
+export interface ITreeVirtualization {
+  viewPort: { from: number; to: number };
+  setRootRef(element: HTMLElement | null): void;
+}
+
+export function useTreeVirtualization(): ITreeVirtualization {
+  const mountOptimization = useObjectRef<IPrivate>(
+    () => ({
+      observer: null,
+      element: null,
+      viewPort: observable({ from: 0, to: 0 }),
+      setRootRef(element: HTMLElement | null) {
+        if (this.element === element) {
+          return;
+        }
+
+        if (this.element) {
+          this.dispose();
+        }
+
+        this.element = element;
+        if (element) {
+          this.observer = new ResizeObserver(this.handleResize);
```

Comment:
this one needs to be tested manually. cause as far as I remember `resizeobserver` can track only one axis. wether is width or height (dont remember which one). if it is important to track both I would recommend to use `addEventListener('resize', this.handleResize)`